File: agent/CloakDlp.Agent/Detection/Ctph.cs
using System.Text;

namespace CloakDlp.Agent.Detection;

// Context-Triggered Piecewise Hashing (CTPH) — a from-scratch fuzzy hash, algorithmically in
// the spirit of ssdeep/TLSH but not binary-compatible with either. Small edits only perturb the
// signature characters near the edit, so similarity between two hashes degrades gracefully
// instead of falling off a cliff the way a cryptographic hash's would.
//
// This exact spec (window size, trigger rule, alphabet, block-size formula) is reimplemented in
// the console backend (console-backend/app/ctph.py) — the two MUST stay in lockstep, or a hash
// produced by one side is meaningless to the other. See ARCHITECTURE.md.
public static class Ctph
{
    private const string Alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    private const int Window = 7;
    private const int MinBlockSize = 3;
    private const int TargetSigLength = 64;
    private const uint FnvOffsetBasis = 0x811C9DC5;
    private const uint FnvPrime = 0x01000193;

    public static string Hash(byte[] data)
    {
        if (data.Length == 0) return $"{MinBlockSize}::";

        var b = SelectBlockSize(data.Length);
        return $"{b}:{Signature(data, b)}:{Signature(data, b * 2)}";
    }

    // 0-100. Only comparable when block sizes match or one is exactly double the other —
    // mirrors ssdeep's own approach of generating two block-size signatures per hash so
    // similarly-sized-but-not-identical documents still have a common ground to compare on.
    public static int Similarity(string hashA, string hashB)
    {
        var (blockA, sigBA, sig2BA) = Parse(hashA);
        var (blockB, sigBB, sig2BB) = Parse(hashB);

        if (blockA == blockB) return EditSimilarity(sigBA, sigBB);
        if (blockA == blockB * 2) return EditSimilarity(sigBA, sig2BB);
        if (blockB == blockA * 2) return EditSimilarity(sig2BA, sigBB);
        return 0;
    }

    private static (int block, string sigB, string sig2B) Parse(string hash)
    {
        var parts = hash.Split(':');
        return (int.Parse(parts[0]), parts.Length > 1 ? parts[1] : "", parts.Length > 2 ? parts[2] : "");
    }

    private static uint Fnv1aStep(uint h, byte b) => unchecked((h ^ b) * FnvPrime);

    private static uint WindowHash(ReadOnlySpan<byte> window)
    {
        var h = FnvOffsetBasis;
        foreach (var b in window) h = Fnv1aStep(h, b);
        return h;
    }

    private static int SelectBlockSize(int length)
    {
        var b = MinBlockSize;
        while ((double)length / b > TargetSigLength) b *= 2;
        return b;
    }

    private static string Signature(byte[] data, int blockSize)
    {
        var sb = new StringBuilder();
        var piece = FnvOffsetBasis;
        var window = new byte[Window];
        var windowLen = 0;

        foreach (var b in data)
        {
            piece = Fnv1aStep(piece, b);

            if (windowLen < Window)
            {
                window[windowLen++] = b;
            }
            else
            {
                Array.Copy(window, 1, window, 0, Window - 1);
                window[Window - 1] = b;
            }

            if (windowLen == Window)
            {
                var h = WindowHash(window);
                if (h % blockSize == blockSize - 1)
                {
                    sb.Append(Alphabet[(int)(piece % 64)]);
                    piece = FnvOffsetBasis;
                }
            }
        }

        if (piece != FnvOffsetBasis || sb.Length == 0)
            sb.Append(Alphabet[(int)(piece % 64)]);

        return sb.ToString();
    }

    private static int EditSimilarity(string a, string b)
    {
        if (a.Length == 0 && b.Length == 0) return 100;
        var distance = LevenshteinDistance(a, b);
        var maxLen = Math.Max(a.Length, b.Length);
        return maxLen == 0 ? 100 : (int)Math.Round(100.0 * (1.0 - (double)distance / maxLen));
    }

    private static int LevenshteinDistance(string a, string b)
    {
        var dp = new int[a.Length + 1, b.Length + 1];
        for (var i = 0; i <= a.Length; i++) dp[i, 0] = i;
        for (var j = 0; j <= b.Length; j++) dp[0, j] = j;

        for (var i = 1; i <= a.Length; i++)
        {
            for (var j = 1; j <= b.Length; j++)
            {
                var cost = a[i - 1] == b[j - 1] ? 0 : 1;
                dp[i, j] = Math.Min(Math.Min(dp[i - 1, j] + 1, dp[i, j - 1] + 1), dp[i - 1, j - 1] + cost);
            }
        }

        return dp[a.Length, b.Length];
    }
}

File: console-backend/app/ctph.py
"""Context-Triggered Piecewise Hashing (CTPH); a from-scratch fuzzy hash, algorithmically in
the spirit of ssdeep/TLSH but not binary-compatible with either. Small edits (insertions,
deletions, reformatting) only perturb the signature characters near the edit, not the whole
hash, so similarity between two hashes degrades gracefully instead of falling off a cliff the
way a cryptographic hash's would.

This exact spec (window size, trigger rule, alphabet, block-size formula) is reimplemented
in the agent (agent/CloakDlp.Agent/Detection/Ctph.cs); the two MUST stay in lockstep, or a
hash produced by one side becomes meaningless to the other. See ARCHITECTURE.md.
"""

ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
WINDOW = 7
MIN_BLOCK_SIZE = 3
TARGET_SIG_LENGTH = 64
FNV_OFFSET_BASIS = 0x811C9DC5
FNV_PRIME = 0x01000193
MASK32 = 0xFFFFFFFF


def _fnv1a_step(h: int, byte: int) -> int:
    return ((h ^ byte) * FNV_PRIME) & MASK32


def _window_hash(window: bytes) -> int:
    h = FNV_OFFSET_BASIS
    for b in window:
        h = _fnv1a_step(h, b)
    return h


def _select_block_size(length: int) -> int:
    b = MIN_BLOCK_SIZE
    while length / b > TARGET_SIG_LENGTH:
        b *= 2
    return b


def _signature(data: bytes, block_size: int) -> str:
    out = []
    piece = FNV_OFFSET_BASIS
    window = bytearray()

    for byte in data:
        piece = _fnv1a_step(piece, byte)

        window.append(byte)
        if len(window) > WINDOW:
            window.pop(0)
        if len(window) == WINDOW:
            h = _window_hash(bytes(window))
            if h % block_size == block_size - 1:
                out.append(ALPHABET[piece % 64])
                piece = FNV_OFFSET_BASIS

    if piece != FNV_OFFSET_BASIS or not out:
        out.append(ALPHABET[piece % 64])

    return "".join(out)


def hash_bytes(data: bytes) -> str:
    """Returns "blockSize:sigAtB:sigAt2B". Callers should discard `data` immediately after -
    this is a one-way fuzzy digest, not a copy of the content."""
    if not data:
        return f"{MIN_BLOCK_SIZE}::"

    b = _select_block_size(len(data))
    return f"{b}:{_signature(data, b)}:{_signature(data, b * 2)}"
